geoAssembler/optimiser/centre.py

```diff
@@ -0,0 +1,180 @@
+import numpy as np
+
+from collections import namedtuple
+from itertools import product
+
+from .utility import Integrator
+
+from extra_geom.detectors import DetectorGeometryBase
+from typing import Union
+
+
+class CentreOptimiser:
+    """
+    Object used to manage the optimisation of the centre position
+    via maximization of the azimuthal integration result.
+
+    The centre position leading to the highest azimuthal integration
+    value at a radial bin should be the one where the diffraction
+    rings form as straight a line as possible, and the rings becoming
+    straight lines in polar coordinates indicates an accurate centre.
+    """
+    def __init__(self, geom: DetectorGeometryBase,
+                 frame: np.ndarray, sample_dist_mm: Union[int, float],
+                 unit: str = "2th_deg"):
+        self.frame = frame
+        self.integrator = Integrator(geom, sample_dist_mm, unit)
+        self.integrate2d = self.integrator.integrate2d
+
+        #  Slightly dodgy way to pull the quadrant corner positions out of geom
+        #  TODO: Suggest adding this in to extra-geom?
+        self.original_centre = [
+            m[0].corners()[0, :2]*geom._get_plot_scale_factor('px')
+            for m
+            in geom.modules
+        ][::4]
+
+    def _loss_function(self, centre_offset):
+        """
+        Simple cost function which computes the 1d azimuthal integration
+        result with a centre offset, it then returns one over this result
+        so that it can be minimised.
+
+        Note: the first and last 100 radial bins are excluded, as these can
+        lead to artefacts which cause the optimisation to fail.
```

Comment:
Are there guaranteed to always be > 200 bins? Is the failure comprehensible if there aren't?

Reply:
Yep, the number of bins comes from this kind of silly bit:  `self.radius = ((self.size[0]/2)**2 + (self.size[1]/2)**2)**(1/2)`, so for assembled AGIPD images it should always be at least 700 ish I expect.